src/Gaffer/ValuePlug.cpp

```diff
@@ -798,10 +890,17 @@ ValuePlug::~ValuePlug()
 	// with a count greater than ours. We do this atomically because
 	// although each graph should only be edited on one thread, it is
 	// permitted to edit multiple graphs in parallel.
-	if( m_dirtyCount != std::numeric_limits<uint64_t>::max() )
+
+	uint64_t nextDirtyCount = m_dirtyCount + 1;
+	if( nextDirtyCount >= DIRTY_COUNT_RANGE_HALF )
+	{
+		nextDirtyCount = 0;
+	}
+
+	if( nextDirtyCount > g_dirtyCountEpoch )
```

Comment:
Doesn't this new version lead to a lot of unnecessary clearing of the cache?

- Make two plugs
- Dirty one N times and then delete it. `g_dirtyCountEpoch == N`
- Delete the second plug. `nextDirtyCount <= g_dirtyCountEpoch` and we clear the cache unnecessarily

I think this scenario is actually very likely. I think the commonest cause of plug deletions is the use of temporary utility plugs like the ones that the UIEditor makes to edit presets, and those will naturally have very low dirty counts.

Am I missing something?

Reply:
Dangit, you're totally right.

I started thinking about how we can detect a wraparound just based on the current values of g_dirtyCountEpoch and m_dirtyCount, but there isn't really enough information.

New proposal for different approach:

ValuePlug::dirty() should reset g_dirtyCountEpoch to 0 when it does a clearCache() - if the cache is being cleared anyway, there's no reason not to reset it.

Instead of replicating any of the logic in dirty, ~ValuePlug should just start by calling dirty().  This will increment m_dirtyCount, taking care of wrapping and potentially clearing.

Then ~ValuePlug can just do:
`while( !g_dirtyCountEpoch.compare_exchange_weak( epoch, std::max( epoch, m_dirtyCount ) ) )`

This is safe because if a wraparound happened, g_dirtyCountEpoch would have been reset by dirty(), so we don't need to worry about the `max` causing epoch to stay stuck high.

Reply:
No, wait, this still doesn't really work great ...

The biggest problem is that it feels like both of us have been thinking about this like the dirty counts on all plugs advance together, so that we catch the point in time where a wrap happens, and then everything gets back to normal.

But in reality, if a plug reaches DIRTY_COUNT_RANGE_HALF, a far more likely scenario is that the dirty counts of the other plugs in the scene are evenly distributed between 0 and DIRTY_COUNT_RANGE_HALF ... meaning cache clears are likely to continue occuring continually for the rest of the session - nothing is ever going to force the dirty counts back into the same portion of the range, once they've spread out.

The ping ponging between updates of plugs that have wrapped and haven't wrapped I think also might cause problems for the approach I just described that would cause excessive cacheClears ... though maybe this isn't that big a problem in itself, now that I've realized you would start getting constant cacheClears anyways without even deleting any plugs?

I'm not sure how we actually solve this without setting all m_dirtyCount on all plugs to 0 whenever we do a cacheClear.